File: day01/src/main.py
from collections import Counter
from pathlib import Path


def parse_input(filename):
    data = Path(filename).read_text().splitlines()
    return zip(*[map(int, line.split()) for line in data])


def part1():
    left_list, right_list = map(sorted, parse_input('my_input.txt'))
    total_distance = sum(abs(l - r) for l, r in zip(left_list, right_list))

    print(f'Answer: {total_distance}')


def part2():
    left_list, right_list = parse_input('my_input.txt')
    right_count = Counter(right_list)
    similarity_score = sum(l * right_count[l] for l in left_list)

    print(f'Answer: {similarity_score}')


if __name__ == '__main__':
    part1()
    part2()

File: day01/src/main.rs
use std::collections::HashMap;
use std::fs;

fn parse_input(filename: &str) -> (Vec<i32>, Vec<i32>) {
    fs::read_to_string(filename)
        .expect("Failed to read file")
        .trim()
        .lines()
        .map(|line| {
            line.split_whitespace()
                .map(|n| n.parse().unwrap())
                .collect::<Vec<i32>>()
        })
        .map(|nums| (nums[0], nums[1]))
        .unzip()
}

fn part1() {
    let total_distance: i32 = {
        let (mut left_list, mut right_list) = parse_input("my_input.txt");
        left_list.sort();
        right_list.sort();
        left_list
            .iter()
            .zip(right_list.iter())
            .map(|(l, r)| (l - r).abs())
            .sum()
    };

    println!("Answer: {}", total_distance);
}

fn part2() {
    let similarity_score: i32 = {
        let (left_list, right_list) = parse_input("my_input.txt");
        let right_count = right_list.into_iter().fold(HashMap::new(), |mut acc, num| {
            *acc.entry(num).or_insert(0) += 1;
            acc
        });
        left_list
            .iter()
            .map(|num| num * right_count.get(num).unwrap_or(&0))
            .sum()
    };

    println!("Answer: {}", similarity_score);
}

fn main() {
    part1();
    part2();
}
